IHP/Breadcrumb/ViewFunctions.hs

```diff
@@ -0,0 +1,45 @@
+module IHP.Breadcrumb.ViewFunctions (
+    module IHP.Breadcrumb.Types,
+    renderBreadcrumb,
+    breadcrumbLink,
+    breadcrumbLinkExternal,
+    breadcrumbText,
+) where
+
+import IHP.Prelude
+import IHP.Breadcrumb.Types
+
+import IHP.ControllerSupport
+
+import Text.Blaze.Html (Html)
+
+import IHP.View.Types (BreadcrumbsView(..), styledBreadcrumb, styledBreadcrumbItem)
+import IHP.ViewSupport (theCSSFramework)
+import IHP.ControllerPrelude
+
+renderBreadcrumb :: (?context :: ControllerContext) => [BreadcrumbItem] -> Html
+renderBreadcrumb breadcrumbItems = styledBreadcrumb theCSSFramework theCSSFramework breadcrumbItems breadcrumbsView
+        where
+            breadcrumbsView = BreadcrumbsView
+                { breadcrumbItems = forEachWithIndex breadcrumbItems (\(index, breadcrumbItem) ->
+                    let
+                        isLast = index == length breadcrumbItems
```

Comment:
I think the issue is here :) Like this should be:

```haskell
isLast = index == (length breadcrumbItems - 1)
```

Reply:
Maybe it makes sense to also add a test for this rendering function? That would have caught the issue earlier

Reply:
Good catch, thanks.

Here's the test I'd like to write

```haskell
it "should render the wrapping breadcrumb and last item as active" do
    let breadcrumbs = [breadcrumbText "First item", breadcrumbText "Last item"]

    -- This below doesn't work
    let frameworkConfig = getFrameworkConfig ?context
    let ?context = frameworkConfig

    renderBreadcrumb breadcrumbs `shouldRenderTo` "..."
```

How can we mock a `?context` in the test?

Reply:
Ok, I found:

`let ?context = createControllerContextWithParams [("page", "1")]`

and

```haskell
context <- newControllerContext
let ?context = context
```

Reply:
@mpscholten I'll appreciate your help here :smiley_cat: I believe we'd need a more generic `let ?context = mockControllerContext` so that we can hand it over in turn into `renderBreadcrumb`

Reply:
The `createControllerContextWithParams` is defined like this:

```haskell
createControllerContextWithParams params =
        let
            requestBody = FormBody { params, files = [] }
            request = Wai.defaultRequest
            requestContext = RequestContext { request, respond = error "respond", requestBody, vault = error "vault", frameworkConfig = error "frameworkConfig" }
        in FrozenControllerContext { requestContext, customFields = TypeMap.empty }
```

I guess in our case the problem is that `theCSSFramework` accesses the `frameworkConfig` field of the context to retrieve the current configured css framework.

With the framework config it could look like this:

```haskell
createControllerContextWithCSSFramework cssFramework =
        let
            frameworkConfig = FrameworkConfig.buildFrameworkConfig do
                    option cssFramework
            requestBody = FormBody { params, files = [] }
            request = Wai.defaultRequest
            requestContext = RequestContext { request, respond = error "respond", requestBody, vault = error "vault", frameworkConfig = frameworkConfig }
        in FrozenControllerContext { requestContext, customFields = TypeMap.empty }
```